KEP-4412: Projected Service Account Tokens for Kubelet Image Credential Providers (#4846)
kubernetes/enhancements: keps/sig-auth/4412-projected-service-account-tokens-for-kubelet-image-credential-providers/README.md

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

By moving to image pulls that are based on KSA tokens, we seek to:

- Reduce secret management overhead for developers and admins
- Reduce security risks associated with long lived secrets

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

- Allow workloads to pull images based on their own runtime identity without long lived / persisted secrets
- Avoid needing a kubelet/node based identity to pull images

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

We will expand the on-disk kubelet credential provider configuration to allow an

> optional `audience` to be configured.  When this field is not set, no KSA

Author: there are more fields, but this field is the required one to enable the feature in addition to the feature flag, so only mentioning that in the proposal section.

   1. Only a single audience can be configured.
   2. The credential provider can be set up multiple times for each audience.
2. `ServiceAccountAnnotations` is the list of annotations on the service account that plugin can use to extract additional information required to fetch credentials. The
annotations will be passed to the plugin along with the image and service account token. The plugin is responsible for validating and using the annotations as needed.
   1. Service account annotations are significantly more stable (and map naturally to the service account being the partioning dimension).
   2. This field makes the provider opt into getting the annotations it wants, avoids sending arbitrary annotation contents down to the plugin (including stuff like client-side apply annotations) and shrinks the set of annotations that could invalidate the cache.

```diff
--- a/pkg/kubelet/apis/config/types.go
+++ b/pkg/kubelet/apis/config/types.go
@@ -653,10 +653,19 @@ type CredentialProvider struct {
        // +optional
        Env []ExecEnvVar
 

> +       // ServiceAccountTokenAudience is the intended audience for the projected service account token.

Author: good idea! I'll make that change.

+       // If set, the kubelet will generate a service account token for the audience and pass it to the plugin.
+       // +optional
+       ServiceAccountTokenAudience string
+

> +       // ServiceAccountAnnotations is the list of annotations to be passed to the plugin.

Reviewer: aren't annotations often key-value pairs? are users passing keys to this and then the kubelet populates the value? or is it only keys, or some combination?

Author: The `ServiceAccountAnnotations` in the credential provider config is a list indicating which annotations from Kubernetes Service Account (KSA) the kubelet needs to send as part of the credential provider request.

For example,

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: my-service-account
  namespace: my-namespace
  annotations:
    domain.io/identity-id: "12345"
    domain.io/identity-type: "user"
    domain.io/annotation-that-will-not-be-passed: "value"
```
```yaml
apiVersion: kubelet.config.k8s.io/v1
kind: CredentialProviderConfig
providers:
  - name: acr-credential-provider
    matchImages:
      - "*.registry.io/*"
    defaultCacheDuration: "10m"
    apiVersion: credentialprovider.kubelet.k8s.io/v1
    serviceAccountTokenAudience: my-audience
    serviceAccountAnnotations:
      - domain.io/identity-id
      - domain.io/identity-type
```

Example credential provider request:

```json
{
  "image": "my-image",
  "serviceAccountToken": "<service-account-token>",
  "serviceAccountAnnotations": {
    "domain.io/identity-id": "12345",
    "domain.io/identity-type": "user"
  }
}
```

I have this documented in the KEP as well. I guess the name `ServiceAccountAnnotations` might be misleading. The intent is to say the field here defines what subset of annotations from KSA need to be plumbed to the plugin. I'm open to suggestions on the field name.

Reviewer: I think I missd it in the kep generally, maybe it could use a summary in this section so it's clear that's the goal. also maybe `ServiceAccountAnnotationKeys` ?

> +       // This field is optional and may be empty. Plugins may use this field to extract

Author: Added a note about it.

+       // or identity type to use when fetching credentials.
+       // Only annotations defined in this list will be passed to the plugin and the plugin is 
+       // responsible for validating the annotations.
+       // +optional
+       ServiceAccountAnnotations []string
```

Example credential provider configuration:

```yaml
apiVersion: kubelet.config.k8s.io/v1
kind: CredentialProviderConfig
providers:
  - name: acr-credential-provider
    matchImages:
      - "*.registry.io/*"
    defaultCacheDuration: "10m"
    apiVersion: credentialprovider.kubelet.k8s.io/v1
    serviceAccountTokenAudience: my-audience
    serviceAccountAnnotations:
      - domain.io/identity-id
      - domain.io/identity-type
```

##### Kubernetes API Server (KAS) Changes

Add a new flag `--image-pull-credential-provider-audiences` to KAS to allow configuring the list of audiences for which the Kubelet is allowed to generate service account tokens for image pulls.

The node audience restriction will be enforced by the [NodeRestriction admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction)

> and kubelet will only be allowed to generate a service account token for the audience configured in the credential provider configuration if it is present in the list of audiences configured in KAS.

Reviewer: If this node audience field is only for image pull creds: how will the KAS differentiate between SA tokens the kubelet requests for other purposes?

Author: This is the list of allowed audiences the kubelet can request a token for. IIUC, the API server doesn't validate _why_ kubelet is requesting a token for but only checks if the audience is referenced somewhere (in this case configured in the API server as a list of audiences)

xref: https://github.com/kubernetes/kubernetes/compare/master...liggitt:kubernetes:node-audience-restriction?w=1

Reviewer: so maybe this flag should be named more generically to reflect that. `--allowed-kubelet-audiences` or something?

Author: `--allowed-kubelet-audiences` is fine! I'll update it.

```diff
--- a/staging/src/k8s.io/kubelet/pkg/apis/credentialprovider/types.go
+++ b/staging/src/k8s.io/kubelet/pkg/apis/credentialprovider/types.go
@@ -32,6 +32,16 @@ type CredentialProviderRequest struct {
        // credential provider plugin request. Plugins may optionally parse the image
        // to extract any information required to fetch credentials.
        Image string
+
+       // serviceAccountToken is the service account token bound to the pod for which
+       // the image is being pulled. This token is only sent to the plugin if the
+       // audience field is configured in the kubelet's credential provider configuration.
+       ServiceAccountToken string
+
+       // serviceAccountAnnotations is a map of annotations on the service account bound to the
+       // pod for which the image is being pulled. The list of annotations in the service account
+       // that need to be passed to the plugin is configured in the kubelet's credential provider
+       // configuration.

> +       ServiceAccountAnnotations map[string]string

Author: `ServiceAccountAnnotationKeys` in credential provider config will only be allowed to be set when the audience field is configured. This will be handled in validation, so we will not end up in a situation where the token is empty but annotations exist.

  annotations:
    domain.io/identity-id: "12345"
    domain.io/identity-type: "user"
    domain.io/annotation-that-will-not-be-passed: "value"
```

Example credential provider request:

```json
{
  "image": "my-image",
  "serviceAccountToken": "<service-account-token>",
  "serviceAccountAnnotations": {
    "domain.io/identity-id": "12345",
    "domain.io/identity-type": "user"
  }
}
```

#### Caching Credentials

The cache key generation will be updated to support new caching modes while preserving the existing behavior when the audience field is not set.

##### Existing behavior

If the `serviceAccountTokenAudience` field is not set in the provider configuration for the plugin, the key generation should fall back to the current behavior: using only the image registry URL or other existing identifiers.

##### New behavior when the `serviceAccountTokenAudience` field is set

Based on the `PluginCacheKeyType` definitions (Image, Registry, Global), here’s a breakdown of how cache keys would be generated:

1. `GlobalPluginCacheKeyType` (Global)
   - Global cache key type will not be allowed to use with the `serviceAccountTokenAudience` field configured for the plugin.

> 2. `RegistryPluginCacheKeyType` (Registry)

Reviewer 2: Which credentials will be used when I am attempting to pull an image with an SA token and the image/registry are already in the cache, but the combination of `image/registry` + SA token isn't?

Author: The cache is maintained per provider, so when the provider is configured with service account audience, the cache will always contain the service account as a dimension for the cache key.

tied to a particular workload while avoiding the need for long lived, persisted secrets.

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

By moving to image pulls that are based on KSA tokens, we seek to:

- Reduce secret management overhead for developers and admins
- Reduce security risks associated with long lived secrets

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

- Allow workloads to pull images based on their own runtime identity without long lived / persisted secrets
- Avoid needing a kubelet/node based identity to pull images

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

> - Passthrough mode where KSA tokens are sent directly to the CRI

Reviewer 3: maybe move from "non-goals" to "possible future work", maybe with a sketch of what it would look like ... not sure we care about making it part of initial graduation criteria for this, but ensuring we don't block expanding in this direction would be ideal

Author: done!

-->

TODO(aramase)
1. add details about the ensure image pulls KEP and impact on this KEP - Talk about how the caching logic will require changes.

#### Credential Provider Configuration

Add `Audience` field to `CredentialProvider` struct in `CredentialProviderConfig`.
This field will be optional and if not set, no KSA token will be sent to the plugin.

- Only a single audience can be configured.
- The credential provider can be set up multiple times for each audience.

```diff
--- a/pkg/kubelet/apis/config/types.go
+++ b/pkg/kubelet/apis/config/types.go
@@ -648,6 +648,11 @@ type CredentialProvider struct {
        // to pass argument to the plugin.
        // +optional
        Env []ExecEnvVar
+
+       // Audience is the intended audience for the credentials. If set, the kubelet will
+       // generate a service account token for the audience and pass it to the plugin.
+       // +optional
+       Audience string
```

For the pod to be able to use the projected service account token for image pulls, the pod spec must include a projected service account token volume with the audience set to the credential provider audience. 

> Only the volume is required in the pod spec, and it does not need to be mounted. The node audience restriction will be enforced by the [NodeRestriction admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction)

Reviewer 3: If the only reason for this is to inform kube-apiserver of the audiences the kubelet should be allowed to request, giving it that information directly (via config/invocation or dedicated API) seems better than a dummy unmounted volume.

Author: Added a new flag for this in KAS

  - name: my-container
    image: my-image-to-pull
  volumes:
  - name: service-account-token
    projected:
      sources:
      - serviceAccountToken:
          path: wi-for-image-pull
          audience: my-audience # must match the credential provider audience
```

Example credential provider configuration:

```yaml
apiVersion: kubelet.config.k8s.io/v1
kind: CredentialProviderConfig
providers:
  - name: acr-credential-provider
    matchImages:
      - "*.registry.io/*"
    defaultCacheDuration: "10m"
    apiVersion: credentialprovider.kubelet.k8s.io/v1
    audience: my-audience
```

#### Credential Provider Request API

Add `ServiceAccountToken` and `PodAnnotations` fields to `CredentialProviderRequest`.

1. `ServiceAccountToken` is the service account token bound to the pod for which the image is being pulled.
    If the `Audience` field is configured in the kubelet's credential provider configuration, the token will be sent to the plugin.
2. `PodAnnotations` is a map of annotations on the pod for which the image is being pulled.
    This field is optional and may be empty. Plugin may use this field to extract additional information required to fetch credentials.

>     For example, a plugin may use annotations to determine the identity ID to use when fetching credentials.

Author: Removed pod annotations and added a new service account annotations field in the credential provider config to define the annotations that'll be plumbed to the plugin as part of the credential provider request.

<!-- Would it be possible for a plugin to talk to kubelet API and get the info instead of passing the pod annotations? -->
<!-- We already send pod annotations to CRI, so this is consistent with that -->

<!-- Add note about a plugin that just returns the service account token without any additional logic and that gets passed to the CRI. -->

```diff
--- a/staging/src/k8s.io/kubelet/pkg/apis/credentialprovider/types.go
+++ b/staging/src/k8s.io/kubelet/pkg/apis/credentialprovider/types.go
@@ -32,6 +32,16 @@ type CredentialProviderRequest struct {
        // credential provider plugin request. Plugins may optionally parse the image
        // to extract any information required to fetch credentials.
        Image string
+
+       // serviceAccountToken is the service account token bound to the pod for which
+       // the image is being pulled. This token is only sent to the plugin if the
+       // audience field is configured in the kubelet's credential provider configuration.
+       ServiceAccountToken string
+
+       // podAnnotations is a map of annotations on the pod for which the image is being pulled.

> +       // This field is optional and may be empty. Plugins may use this field to extract

Reviewer 4: What about service account annotations?

+       // additional information required to fetch credentials.
+       PodAnnotations map[string]string
 }
```

#### Credential Provider Response API

> When the `audience` field is set in the kubelet's credential provider configuration, the plugin must return a `CredentialProviderResponse` with the `cacheMode` field set to `ServiceAccount` or `None`. If the cache mode is not set, kubelet will error out and not use the response. The field is required and serves as a marker to indicate that the plugin understands the new fields in the request.

Reviewer 3: I think it's questionable to send service account information (either dimensions or an actual token) to a credential provider and make it possible for it to return a result that is cached across service accounts

+       // There are two valid values for the cache mode: ServiceAccount and None.
+       // If an invalid value is specified, the response will NOT be used by the kubelet.
+       CacheMode PluginCacheMode
```

#### Caching Credentials

The cache key generation will be updated to support new caching modes while preserving the existing behavior when the audience field is not set.

##### Existing behavior

If the `audience` field is not set in the provider configuration for the plugin, the key generation should fall back to the current behavior: using only the image registry URL or other existing identifiers.

##### New behavior when the `audience` field is set

Based on the `PluginCacheKeyType` definitions (Image, Registry, Global) and the newly introduced cache modes (ServiceAccount, None), here’s a breakdown of how cache keys would be generated and whether cache mode impacts their behavior:

1. `GlobalPluginCacheKeyType` (Global)
   - cache mode does not affect the cache key generation since it is designed to provide credentials for all images.
2. `RegistryPluginCacheKeyType` (Registry)
   - `ServiceAccount` cache mode:

>      - The cache key will be generated based on the image registry URL, service account token bound to the pod (namespace, name, UID), hash of the pod annotations.

Reviewer 3: pod annotations are questionable as a cache key... admission and controllers put weird random values in those that differ across replicas

service account annotations would be significantly more stable (and would map much more naturally to the service account being the partitioning dimension), though I don't love pushing opaque user-controlled values down to providers

making the credential provider opt into getting the annotations (maybe even the specific annotations it wants) avoids sending arbitrary annotation content down to it (including stuff like client-side apply annotations), and shrinks the set of annotations that could invalidate caching if they change

that any pod running on that node can access those images).  A pod should instead be able
to use its own identity to pull images, i.e. we should enable image pull authorization to be
tied to a particular workload while avoiding the need for long lived, persisted secrets.

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

By moving to image pulls that are based on KSA tokens, we seek to:

- Reduce secret management overhead for developers and admins
- Reduce security risks associated with long lived secrets

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

- Allow workloads to pull images based on their own runtime identity without long lived / persisted secrets
- Avoid needing a kubelet/node based identity to pull images

> ### Non-Goals

Reviewer 5: Direct passthrough support is a non-goal right because we don't know what level of flexibility is need for the username?

Author: I moved it from `Non-Goals` to `Possible future work` as per https://github.com/kubernetes/enhancements/pull/4846#discussion_r1757308948

Reviewer 5: Okay SGTM

This change reduces the attack surface and the risk of credential leaks, aligning with
best practices for security.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

To keep this KEP small both in terms of implementation and scope,
the design is written to meet the following criteria:

- No changes to any Kubernetes REST APIs
- No changes to the CRI API
- No changes to how the Kubelet interacts with the CRI API
- No changes to any registry
- Minimal changes to Kubelet
- Minimal changes to Kubelet credential providers

> 

Reviewer 5: ```suggestion
- Minimal changes to Kube API server

```

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

#### Credential Provider Configuration

Adding a new field `TokenAttributes` to the `CredentialProvider` struct in the kubelet configuration.

1. `ServiceAccountTokenAudience` is the intended audience for the credentials. If set, the kubelet will generate a service account token for the audience and pass it to the plugin.
   1. Only a single audience can be configured.

>    2. The credential provider can be set up multiple times for each audience.

Reviewer 5: I think this requires the use of sym links to give each provider a different name?

   3. By setting this field, the credential provider is opting into using service account tokens for image pull.

> 2. `ServiceAccountAnnotationKeys` is the list of annotation keys that the pllugin is interested in. The keys defined in this list will be extracted from the corresponding service account

Reviewer: s/pllugin/plugin

>    and passed to the plugin as part of `CredentialProviderRequest`. Plugins can use to extract additional information required to fetch credentials. The plugin is responsible for validating the existence of annotations and their values.

Reviewer 5: Describe that many existing integrations in this space require extra metadata to link the KSA name to a cloud identity.  Also note that no existing integrations use this metadata for any form of security decision, it is simply to aid in doing token exchanges with the KSA token.

```diff
--- a/pkg/kubelet/apis/config/types.go
+++ b/pkg/kubelet/apis/config/types.go
@@ -653,10 +653,19 @@ type CredentialProvider struct {
        // +optional
        Env []ExecEnvVar
 
+       // TokenAttributes is the configuration for the service account token that will be passed to the plugin.
+       // The credential provider opts in to using service account tokens for image pull by setting this field.
+       // When this field is set, kubelet will generate a service account token bound to the pod for which the
+       // image is being pulled and pass to the plugin as part of CredentialProviderRequest along with other
+       // attributes required by the plugin.
+       //
+       // The service account metadata and token attributes will be used as a dimension to cache
+       // the credentials in kubelet. The cache key is generated by combining the service account metadata

> +       // (namespace, name, UID, and annotations map with keys defined in ServiceAccountTokenAttribute.serviceAccountAnnotationKeys).

Reviewer 5: I assume the values associated with those keys will also be part of the cache key?

Author: that's correct! I had used "annotations map" to indicate but will clarify to make it more clear.

> +       TokenAttributes *ServiceAccountTokenAttribute

Reviewer: pedantic nit: it's odd to me the name of the variable is plural but the type is singular (and not a list). consider choosing plural or singular for both maybe?

Author: thank you, fixed!

+       // the existence of annotations and their values.
+       // This field is optional and may be empty. Plugins may use this field to extract
+       // additional information required to fetch credentials.
        // +optional
+       ServiceAccountAnnotationKeys []string
 }

```

Example credential provider configuration:

```yaml
apiVersion: kubelet.config.k8s.io/v1
kind: CredentialProviderConfig
providers:
  - name: acr-credential-provider
    matchImages:
      - "*.registry.io/*"
    defaultCacheDuration: "10m"
    apiVersion: credentialprovider.kubelet.k8s.io/v1
    tokenAttributes:
      serviceAccountTokenAudience: my-audience
      # list of annotations that the plugin is interested in
      # the annotation key and corresponding value in the service account will be passed to the plugin
      serviceAccountAnnotationKeys:
      - domain.io/identity-id
      - domain.io/identity-type
```

##### Kubernetes API Server (KAS) Changes

Add a new flag `--allowed-kubelet-audiences` to KAS to allow configuring the list of audiences for which the Kubelet is allowed to generate service account tokens for image pulls.

> The node audience restriction will be enforced by the [NodeRestriction admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction)

Reviewer 5: Mention that today we only verify the SA and allow all audiences.  We want to verify the audiences as well, so we need to be explicit about them here because the other sources of audiences can be observed via the kube API, but this one cannot.

```diff
--- a/staging/src/k8s.io/kubelet/pkg/apis/credentialprovider/types.go
+++ b/staging/src/k8s.io/kubelet/pkg/apis/credentialprovider/types.go
@@ -32,6 +32,16 @@ type CredentialProviderRequest struct {
        // credential provider plugin request. Plugins may optionally parse the image
        // to extract any information required to fetch credentials.
        Image string
+
+       // serviceAccountToken is the service account token bound to the pod for which
+       // the image is being pulled. This token is only sent to the plugin if the
+       // tokenAttributes.serviceAccountTokenAudience field is configured in the kubelet's credential provider configuration.
+       ServiceAccountToken string
+
+       // serviceAccountAnnotations is a map of annotations on the service account bound to the
+       // pod for which the image is being pulled. The list of annotations in the service account
+       // that need to be passed to the plugin is configured in the kubelet's credential provider
+       // configuration.
+       ServiceAccountAnnotations map[string]string
 }
```

> Example Kubernetes Service Account:

Reviewer 5: Expand this example to show what happens when a credential provider requests an annotation key that is not set on the KSA.

Author: added a note about it in the credential provider request example.

##### New behavior when the `serviceAccountTokenAudience` field is set

Based on the `PluginCacheKeyType` definitions (Image, Registry, Global), here’s a breakdown of how cache keys would be generated:

1. `GlobalPluginCacheKeyType` (Global)
   - Global cache key type will not be allowed to use with the `tokenAttributes.serviceAccountTokenAudience` field configured for the plugin.
2. `RegistryPluginCacheKeyType` (Registry)
   - The cache key will be generated based on the image registry URL, service account metadata (namespace, name, UID), and hash of the service account annotations that are passed to the plugin in the `CredentialProviderRequest.ServiceAccountAnnotations` field.
3. `ImagePluginCacheKeyType` (Image)
    - The cache key will be generated based on the image URL, service account metadata (namespace, name, UID), and hash of the service account annotations that are passed to the plugin in the `CredentialProviderRequest.ServiceAccountAnnotations` field.

#### How will this work with the Ensure secret pull images KEP?

> [2535-ensure-secret-pull-images](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2535-ensure-secret-pulled-images) is a KEP that aims to give admin the ability to ensure pods that use a image are authorized to access the image. The KEP doesn't factor in the use of service account tokens for image pull. As part of the alpha implementation the 2 features together will not provide a desired outcome. Post alpha, we'll need to update the Ensure secret pull images KEP to factor in the use of service account tokens for image pull.

Reviewer 5: Until the two implementations are updated to work together, I would expect the alpha implementation of this KEP to use the KSA token based flow unless the pod is using image pull policy set to always.  This keeps the feature from misbehaving until we fix the implementations.

Author: updated it!

> Notes from reviewing the current KEP and discussion with @stalz:

Reviewer 2: stalz would've been a much-easier-to-pronounce handle now that I see it :slightly_smiling_face:

1. Different KSA for same image should result in image pull from registry.
2. Same KSA for the image will result in allowing pod to use the image.
   1. How would expiry work in this scenario?

>       1. I think it'll just be tied to the KSA and not so much to the expiry of the token because we're doing the same thing with image pull secrets (considered valid until deleted and recreated). Any change to KSA will result in change in UID at the minimum and that'll result in KSA not found in cache for the image (assuming the key used to store in cache is consistent with the cache key used in the credential provider cache that takes UID into consideration). Need to share the cache key generation logic to be consistent.

Reviewer 2: A change to a UID of an SA means it got deleted and recreated, thus making it a completely different SA, but simple updates to an SA won't cause it to change UIDs. The UID in the cache key protects the system from attempts on name clashes.

Author: Deletion and recreation of SA with the same name in the namespace is what I meant here, I'll clarify it.

enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

This feature is feature gated so explicit opt-in is required on upgrade and explicit opt-out is required on downgrade.

### Version Skew Strategy

<!--
If applicable, how will the component handle version skew with other
components? What are the guarantees? Make sure this is in the test plan.

Consider the following in developing a version skew strategy for this
enhancement:
- Does this enhancement involve coordinating behavior in the control plane and nodes?
- How does an n-3 kubelet or kube-proxy without this feature available behave when this feature is used?
- How does an n-1 kube-controller-manager or kube-scheduler without this feature available behave when this feature is used?
- Will any other components on the node change? For example, changes to CSI,
  CRI or CNI may require updating that component before the kubelet.
-->

> Not applicable because this feature is contained to only the kubelet and does not require communication to other components.

Reviewer 5: What about skew between kubelet and old/new credential plugins?  How does a kubelet admin migrate to the new approach, one workload at a time?  I assume all kubelets would need to be updated with new config/plugins and then restarted, followed by upgrades to workloads?

Also API server config needs to be updated because kubelets can safely use this feature.

Author: updated this section with the details

-->

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: `ServiceAccountTokenForKubeletCredentialProviders`
  - Components depending on the feature gate: kubelet

```go
FeatureSpec{
  Default: false,
  LockToDefault: false,
  PreRelease: featuregate.Alpha,
}
```

###### Does enabling the feature change any default behavior?

<!--
Any change of default behavior may be surprising to users or break existing
automations, so be extremely careful here.
-->

No. After enabling the feature flag in the kubelet, the credential provider still needs to opt into using service account tokens for image pull
by setting the `TokenAttributes` field in the kubelet credential provider configuration.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

<!--
Describe the consequences on existing workloads (e.g., if this is a runtime
feature, can it break the existing applications?).

Feature gates are typically disabled by setting the flag to `false` and
restarting the component. No other changes should be necessary to disable the
feature.

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->

> Yes. The feature flag needs to be disabled and the credential provider configuration for the provider that is using service account tokens for image pull

Reviewer 5: What happens to the cache when the feature is disabled?  What about images that were already pulled?

Author: IIUC disabling the feature gate will result in kubelet on the node being restarted? I looked at https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ and didn't see a dynamic reload option, so I'm assuming it always restarts as part of disabling the feature. The cache is in-memory, so that'll be invalidated. For images that have already been pulled, the behavior will be same as today with any pod being able to access the image. With the ensure image pulls KEP, we will be able to force reauthentication if the image is stored in the `pulled` cache on disk. 

LMK what you think!

Reviewer 5: Maybe mention that kubelet restart will remove the in-memory cache?

Author: done!

and creating new ones, as well as about cluster-level services (e.g. DNS):
  - [Dependency name]
    - Usage description:
      - Impact of its outage on the feature:
      - Impact of its degraded performance or high-error rates on the feature:
-->

### Scalability

<!--
For alpha, this section is encouraged: reviewers should consider these questions
and attempt to answer them.

For beta, this section is required: reviewers must answer these questions.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### Will enabling / using this feature result in any new API calls?

<!--
Describe them, providing:
  - API call type (e.g. PATCH pods)
  - estimated throughput
  - originating component(s) (e.g. Kubelet, Feature-X-controller)
Focusing mostly on:
  - components listing and/or watching resources they didn't before
  - API calls that may be triggered by changes of some Kubernetes resources
    (e.g. update of object X triggers new updates of object Y)
  - periodic API calls to reconcile state (e.g. periodic fetching state,
    heartbeats, leader election, etc.)
-->

> Yes. Kubelet will request a KSA token from KAS for pod + KSA for every provider (audience) that is configured to use service account tokens for image pull.

Reviewer 5: Mention that this load is small because it only happens once on pod startup.

###### Will enabling / using this feature result in introducing new API types?

<!--
Describe them, providing:
  - API type
  - Supported number of objects per cluster
  - Supported number of objects per namespace (for namespace-scoped objects)
-->

No.

###### Will enabling / using this feature result in any new calls to the cloud provider?

<!--
Describe them, providing:
  - Which API(s):
  - Estimated increase:
-->

> No.

Reviewer 5: It could depending on the plugin?

- Reduce secret management overhead for developers and admins
- Reduce security risks associated with long lived secrets

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

- Allow workloads to pull images based on their own runtime identity without long lived / persisted secrets
- Avoid needing a kubelet/node based identity to pull images

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

We will expand the on-disk kubelet credential provider configuration to allow an
optional `tokenAttribute` field to be configured.  When this field is not set, no KSA
token will be sent to the plugin.  When it is set, the Kubelet will provision

> a token with the given audience bound to the current pod and its service

Reviewer 6: will the token be time-bound? If so - how long?

Reviewer 6: will it work nicely with things like stargz snapshotter where secret can be stored for a long time like: https://github.com/containerd/stargz-snapshotter/blob/main/docs/overview.md#cri-based-authentication? 

cc: @samuelkarp

Author: > will the token be time-bound? If so - how long?

Yes, the token will be time-bound. The duration will be fixed at 1h (?). If the cloud provider associates this token duration to the credential lifetime, we don't want it to be too short-lived. 1h seems reasonable to cover those scenarios to prevent frequent calls to cloud provider for token exchange.

Author: > will it work nicely with things like stargz snapshotter where secret can be stored for a long time like: https://github.com/containerd/stargz-snapshotter/blob/main/docs/overview.md#cri-based-authentication?

IIUC, the stargz snapshotter in `CRI-based authentication` gets the same registry credentials as the CRI from kubelet, so that wouldn't change here? The credentials could now be short-lived if the lifetime is tied to the KSA token lifetime but otherwise should work the same way. Please correct me if I'm wrong.

Reviewer 6: stargz is not downloading the whole image in one go. So it may keep streaming pieces after `1h`. So it will end up with the stale credentials

Author: > stargz is not downloading the whole image in one go. So it may keep streaming pieces after 1h. So it will end up with the stale credentials

The credential provider plugin gets the PSAT and exchanges that for a registry credential. The registry credential is the one that `stargz` uses and it's not the PSAT right?

Reviewer: correct. I think this can be compared to current behavior without specific PSAT: if a node wide credential expires while the image is being pulled with stargz, what does the runtime do? I'd expect the same here.

Reviewer 6: oh, ok. So we are not giving time limited token to runtime. If this is the case, it will work

Reviewer 6: I wonder if there is any test we can write that will validate that and ensure we are not breaking this scenario?

Author: The tests in kubelet credential provider wouldn't validate this because it fetches the registry credentials and passes it to CRI (no change in behavior). 

Tests in stargz would be good to confirm it works as expected when the registry credentials expire/it updates its cache when the credentials for the image change.
@SergeyKanzhelev do you know who can confirm if there are tests for this in `stargz`?

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

To keep this KEP small both in terms of implementation and scope,
the design is written to meet the following criteria:

- No changes to any Kubernetes REST APIs
- No changes to the CRI API
- No changes to how the Kubelet interacts with the CRI API
- No changes to any registry
- Minimal changes to Kubelet
- Minimal changes to Kubelet credential providers
- Minimal changes to Kube API Server

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

#### Credential Provider Configuration

Adding a new field `TokenAttributes` to the `CredentialProvider` struct in the kubelet configuration.

> 1. `ServiceAccountTokenAudience` is the intended audience for the credentials. If set, the kubelet will generate a service account token for the audience and pass it to the plugin.

Reviewer 6: before we never sent any secrets to the credential provider. So it was never required to be secured strictly to avoid the attack when it is replaced with some other binary. With this change we make this attack meaningful. Do we have enough documentation suggesting how to secure the cred provider so it cannot be replaced?

Author: I don't see any documentation [here](https://kubernetes.io/docs/tasks/administer-cluster/kubelet-credential-provider/) but we can certainly get that added as part of the implementation.

Reviewer 5: The credential provider is executed as the same user as the kubelet, so probably root, meaning it can trivially see the kubelet creds from the filesystem and have access to anything that the kubelet does (including tokens for pods).

- Deprecate the flag
-->

#### Alpha

- Feature implemented behind a feature flag
- Unit tests for config validation
- Unit tests for current credential provider logic unchanged when token attributes are not set
- Unit tests for credential provider logic when token attributes are set
- Initial e2e tests completed and enabled

#### Post Alpha

- Make sure the feature is compatible with the [Ensure secret pull images KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2535-ensure-secret-pulled-images).

### Upgrade / Downgrade Strategy

<!--
If applicable, how will the component be upgraded and downgraded? Make sure
this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

This feature is feature gated so explicit opt-in is required on upgrade and explicit opt-out is required on downgrade.

> ### Version Skew Strategy

Reviewer 7: I think it works, but please explicitly describe the case of new kubelet contacting old kube-apiserver to request a KSA token with a configured audience.  Old kube-apiserver must reject in this case.

-->

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: `ServiceAccountTokenForKubeletCredentialProviders`
  - Components depending on the feature gate: kubelet

```go
FeatureSpec{
  Default: false,
  LockToDefault: false,
  PreRelease: featuregate.Alpha,
}
```

###### Does enabling the feature change any default behavior?

<!--
Any change of default behavior may be surprising to users or break existing
automations, so be extremely careful here.
-->

No. After enabling the feature flag in the kubelet, the credential provider still needs to opt into using service account tokens for image pull
by setting the `TokenAttributes` field in the kubelet credential provider configuration.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

<!--
Describe the consequences on existing workloads (e.g., if this is a runtime
feature, can it break the existing applications?).

Feature gates are typically disabled by setting the flag to `false` and
restarting the component. No other changes should be necessary to disable the
feature.

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->

> Yes. The feature flag needs to be disabled and the credential provider configuration for the provider that is using service account tokens for image pull

Reviewer 7: What order do the components need to be reconfigured and restarted?  

1. update kubelet configuration to avoid using new credential provider
2. disable the featuregate
3. restart all kubelets
4. restart kube-apiserver

?

Describe the flow here.

and operation of this feature.
Recall that end users cannot usually observe component logs or access metrics.
-->

- [ ] Events
  - Event Reason: 
- [ ] API .status
  - Condition name: 
  - Other field: 
- [ ] Other (treat as last resort)
  - Details:

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

<!--
This is your opportunity to define what "normal" quality of service looks like
for a feature.

It's impossible to provide comprehensive guidance, but at the very
high level (needs more precise definitions) those may be things like:
  - per-day percentage of API calls finishing with 5XX errors <= 1%
  - 99% percentile over day of absolute value from (job creation time minus expected
    job creation time) for cron job <= 10%
  - 99.9% of /health requests per day finish with 200 code

These goals will help you determine what you need to measure (SLIs) in the next
question.
-->

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

<!--
Pick one more of these and delete the rest.
-->

> - [ ] Metrics

Reviewer 7: not required for alpha, but I strongly suggest considering what metrics are useful as you make your first pass through.

Author: sounds good! I've added information about existing metrics that can be used to observe errors. I'll update the KEP as metrics added in alpha or post-alpha.

Notes from reviewing the current KEP and discussion with @stlaz:

1. Different KSA for same image should result in image pull from registry.
2. Same KSA for the image will result in allowing pod to use the image.
   1. How would expiry work in this scenario?
      1. I think it'll just be tied to the KSA and not so much to the expiry of the token because we're doing the same thing with image pull secrets (considered valid until deleted and recreated). Deletion and recreation of KSA will result in change in UID and that'll result in KSA not found in cache for the image (assuming the key used to store in cache is consistent with the cache key used in the credential provider cache that takes UID into consideration). Need to share the cache key generation logic to be consistent.
3. Need an update to `ImagePullCredentials` struct to also store coordinates of the KSA.

> Until the two implementations are updated to work together, the alpha implementation of this KEP will use the KSA token based flow unless the pod is using image pull policy set to `Always`. This keeps the feature from misbehaving until we fix the implementations.

Reviewer 5: ?

```suggestion
Until the two implementations are updated to work together, the alpha implementation of this KEP will use the KSA token based flow only when the pod is using image pull policy set to `Always`. This keeps the feature from misbehaving until we fix the implementations.
```

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

**For non-optional features moving to GA, the graduation criteria must include
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

#### Alpha

- Feature implemented behind a feature flag
- Unit tests for config validation
- Unit tests for current credential provider logic unchanged when token attributes are not set
- Unit tests for credential provider logic when token attributes are set
- Initial e2e tests completed and enabled

#### Post Alpha

- Make sure the feature is compatible with the [Ensure secret pull images KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2535-ensure-secret-pulled-images).

#### Beta

> - The implementation works well with the Ensure secret pull images KEP and supports pod image pull policy set to `Always`

Author: Ah I meant to say supporting the explicit scenario that's unsupported in alpha. The suggest edit clarifies that better.

- How does an n-1 kube-controller-manager or kube-scheduler without this feature available behave when this feature is used?
- Will any other components on the node change? For example, changes to CSI,
  CRI or CNI may require updating that component before the kubelet.
-->

To migrate to the new approach,

1. KAS will need to be updated to the new version to configure the list of audiences for which the kubelet is allowed to generate service account tokens for image pulls via the `--allowed-kubelet-audiences` flag.
2. kubelet on the node needs to be updated to enable the feature flag and configure the credential provider to use service account tokens for image pull via the `TokenAttributes` field in the kubelet credential provider configuration.
   1. The credential provider plugin will need to be updated to use the service account token for the audience configured in the kubelet credential provider configuration.

Migration of the workloads to the new approach can be done per image or per registry basis by configuring the credential provider multiple times with different names for the same plugin (w or w/o service account tokens, different audiences).

When things can fail:

> 1. If the kubelet is updated to enable the feature flag and the credential provider is configured with the `TokenAttributes` field set, but the KAS is not updated with `--allowed-kubelet-audiences` to allow the kubelet to generate service account tokens for the audience configured in the kubelet credential provider configuration, the image pull will fail. The old KAS will reject the token request from the kubelet.

Reviewer 7: > The old KAS will reject the token request from the kubelet.

It's important that this will actually fail instead of being allowed for the safety of this feature on downgrade and disable, right?  Will you manually test this to be sure?  I'm concerned because I don't remember this failing closed today.

Reviewer 7: > We can make is so that the audience validation is beta/enabled by default at least one release before the rest of the functionality reaches beta/enabled by default. WDYT?

That works for me.  Let's get this clarified in the KEP beta criteria please so we don't forget.  I think that means two gates.

Author: Added a new feature gate and updated the beta criteria to include what @enj mentioned.

Notes from reviewing the current KEP and discussion with @stlaz:

1. Different KSA for same image should result in image pull from registry.
2. Same KSA for the image will result in allowing pod to use the image.
   1. How would expiry work in this scenario?
      1. I think it'll just be tied to the KSA and not so much to the expiry of the token because we're doing the same thing with image pull secrets (considered valid until deleted and recreated). Deletion and recreation of KSA will result in change in UID and that'll result in KSA not found in cache for the image (assuming the key used to store in cache is consistent with the cache key used in the credential provider cache that takes UID into consideration). Need to share the cache key generation logic to be consistent.
3. Need an update to `ImagePullCredentials` struct to also store coordinates of the KSA.

> Until the two implementations are updated to work together, the alpha implementation of this KEP will use the KSA token based flow only the pod is using image pull policy set to `Always`. This keeps the feature from misbehaving until we fix the implementations.

Reviewer 5: typo

```suggestion
Until the two implementations are updated to work together, the alpha implementation of this KEP will use the KSA token based flow only when the pod is using image pull policy set to `Always`. This keeps the feature from misbehaving until we fix the implementations.
```

#### Deprecation

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

#### Alpha

- Feature implemented behind a feature flag
- Unit tests for config validation
- Unit tests for current credential provider logic unchanged when token attributes are not set
- Unit tests for credential provider logic when token attributes are set
- Initial e2e tests completed and enabled
- `ServiceAccountNodeAudienceRestriction` feature gate implemented in KAS as a beta feature
  - Audience validation is enabled by default for service account tokens requested by the kubelet

#### Post Alpha

- Make sure the feature is compatible with the [Ensure secret pull images KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2535-ensure-secret-pulled-images).

#### Beta

- The implementation works well with the Ensure secret pull images KEP and supports pod image pull policy set to any value.

> - `ServiceAccountNodeAudienceRestriction` feature gate is beta in KAS and enabled by default. This feature needs to be enabled be beta/enabled by default at least one release before this KEP goes to beta. This is critical to support downgrade use cases.

Reviewer 5: ```suggestion
- `ServiceAccountNodeAudienceRestriction` feature gate is beta in KAS and enabled by default. This feature needs to be beta/enabled by default at least one release before this KEP goes to beta. This is critical to support downgrade use cases.
```
